### PR TITLE
many:  make access to the device model assertion etc contextual via a DeviceCtx hook/DeviceContext interface

### PR DESCRIPTION
**Pull request by pedronis**
This makes the the access mainly in snapstate to device model assertion etc contextual via a DeviceCtx hook/DeviceContext interface. For example it then can be made contextual to a remodeling change/change preparation. The implementation of the hook is provided by devicestate as before it was providing a simpler Model accessor.

This starts with a fix, in the recent reorg InstallMany stopped reusing Install but that way it lost it's check for model/seeded.

I moved out the one check needing model from doInstall because in the end it seemed to force touching more places than seemed reasonable and provoking slightly unexpected tests needing changes.

A couple of places now needs a device model assertion to be set in a more upfront way.

This introduces also a snapstatetest package that helps mocking the shared state mediated by DeviceCtx.

In general a different device model needs a different store, a follow up will switch relevant places using snapstate.Store directly to get one via DeviceContexts too.

Most of the new accessors/helpers going through a DeviceContext support taking a provided one that then they will just use or return, this is not used yet but will alllow this pattern:

    func Install(st *state.State, name, channel string, revision snap.Revision, userID int, flags Flags, 
                   providedDeviceCtx DeviceContext) (*state.TaskSet, error) {

            ...

            // need to have a model set before trying to talk the store
            deviceCtx, err := DevicePastSeeding(st, providedDeviceCtx)
            if err != nil {
                    return nil, er
            }

### CODE REVIEW

File: overlord/snapstate/snapstate.go [hunk 1]
```diff
@@ -75,30 +74,17 @@ func isParallelInstallable(snapsup *SnapSetup) error {
 }
 
 func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int, fromChange string) (*state.TaskSet, error) {
+	// NB: ideally deviceCtx should not need to propagate here!
```
Comment (mvo5) [inline overlord/snapstate/snapstate.go]: Not sure I get this comment - I don't see deviceCtx used in this function?

File: overlord/snapstate/snapstate.go [hunk 2]
```diff
@@ -527,7 +513,21 @@ func validateFeatureFlags(st *state.State, info *snap.Info) error {
 	return nil
 }
 
-func checkInstallPreconditions(st *state.State, info *snap.Info, flags Flags, snapst *SnapState) error {
+func checkInstallPreconditions(st *state.State, info *snap.Info, flags Flags, snapst *SnapState, deviceCtx DeviceContext) error {
```
Comment (mvo5) [inline overlord/snapstate/snapstate.go]: Nice!

File: overlord/snapstate/snapstate.go [hunk 3]
```diff
@@ -837,10 +850,10 @@ func updateManyFiltered(ctx context.Context, st *state.State, names []string, us
 
 	}
 
-	return doUpdate(ctx, st, names, updates, params, userID, flags, fromChange)
+	return doUpdate(ctx, st, names, updates, params, userID, flags, deviceCtx, fromChange)
```
Comment (mvo5) [inline overlord/snapstate/snapstate.go]: Not a blocker or anything but doUpdate() becomes a bit unwieldy.

File: overlord/snapstate/snapstate_test.go [hunk 4]
```diff
@@ -1903,6 +1912,20 @@ func (s *snapmgrTestSuite) TestInstallStateConflict(c *C) {
 	c.Assert(err, ErrorMatches, `snap "some-snap" has changes in progress`)
 }
 
+func (s *snapmgrTestSuite) TestInstallPathTooEarly(c *C) {
```
Comment (mvo5) [inline overlord/snapstate/snapstate_test.go]: Thanks for adding this (and the other) extra test!

File: overlord/snapstate/snapstate.go (hunk 1 above)
Comment (pedronis) [reply overlord/snapstate/snapstate.go]: Would:

// NB:  we should strive not to need or propagate deviceCtx here,  the resulting effects/changes were not pleasant at one point

be clearer?

**Comment by pedronis** [comment]
one spread test failure is real, working on it

**Pull request merged**